### ISSUE TITLE
Run fails with seg fault (invalid memory reference)

### ISSUE THREAD

**Issue by tovogt** [clawpack/geoclaw#528]
While tracing down issue #525, I found a GeoClaw setup that crashes due to a segmentation fault on our cluster computer with 16 OpenMP threads, but doesn't crash on my local desktop computer. I'm running GeoClaw a lot on our cluster and never had a segmentation fault before. Here is the terminal output: [stdout.txt](https://github.com/clawpack/geoclaw/files/7324009/stdout.txt)

Have you ever seen something like this? 

Maybe it's related to the following AMR settings:
```
7                    =: amr_levels_max      
2 2 2 2 2 4          =: refinement_ratios_x 
```
I was adding the seventh AMR level when the error occurred for the first time. When using only 6 levels, the run will not fail.

**Comment by mandli** [comment]
Hmmm.  Well we will have to see what we find might be going on with gfortran then.

**Comment by tovogt** [comment]
Since the last activity in this issue, I mostly continued to work with gfortran. But yesterday, I started a whole bunch of runs with the `ifort` compiler (tested both version 17 and 19, with `ulimit -s unlimited` and `OMP_STACKSIZE=500M`) and found that a lot of those were failing due to segmentation faults. Apparently, the `ifort` compiler is also affected by this issue, but it fails at a different point and for different input data. The same run will fail at a different point in the simulation when run with `gfortran`! 

However, I don't know how to debug with `ifort`: I get the error message `forrtl: severe (174): SIGSEGV, segmentation fault occurred`. With version 19 there is no further output, with version 17, it reports:
```
*** Error in `$WORKDIR/xgeoclaw': double free or corruption (fasttop): 0x0000000001525010 ***
```
Adding the compiler flags `-g` and `-backtrace` doesn't make the output more verbose. Do you know any way how to obtain a more verbose error output from `ifort` in this case?

To be honest, I'm a bit puzzled by this: The topography I provide to GeoClaw has a resolution of 30 arc-seconds (~1 km) and the mesh resolution with maximum refinement is 7 arc-seconds (200 m). In the literature, in the examples and from what I hear from other GeoClaw users, a lot of them use much higher resolution. How is it possible that I still run into these problems while apparently it works for other people?

**Comment by mandli** [comment]
My guess is this is not a problem with topography resolution but perhaps a bug although the fact that you see it this often is a bit puzzling to me as well.  My best guess without looking into it with @mjberger further is that there is something about your location that is causing the problem but who knows.  If you can send us your setup as @mjberger suggested we can try and figure out what's going wrong hopefully.

**Comment by tovogt** [comment]
I described the setup in https://github.com/clawpack/geoclaw/issues/528#issuecomment-942396358 and provided the rundata files. Do you need any other information from my side?

**Comment by mandli** [comment]
Sorry, forgot about that.  We will try and take a look and see if we can reproduce and debug the problem.

**Comment by mjberger** [comment]
HI Thomas,

I've started looking at your issue.  I've duplicated the error with the gfortran-8 compiler. Interestingly it runs to completion with -g, and only dies with optimization -O0 on.  Makes it harder to debug but will give it a try.

Anyhow, 2 things.

1.  Your zip file had an empty  setrun.py.  It would make it easier to have one, e.g. so I could more easily checkpoint before the error, etc.

2.  Why don't you send me your direct email address , so we don't have to use the clawpack notifications.
Mine is ***@***.*** ***@***.***>

Best,
 — Marsha

**Comment by tovogt** [comment]
After stress testing this with a whole batch of jobs, I only have a single run that fails with a segfault (with gfortran). But the error message changed, so it might be completely unrelated to this issue:
```
Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.

Backtrace for this error:
#0  0x2ae8cc720fdf in ???
#1  0x5173ca in gfixup_
	at $CLAW/geoclaw/src/2d/shallow/gfixup.f:234
#2  0x49690d in regrid_
	at $CLAW/amrclaw/src/2d/regrid.f:71
#3  0x50f081 in tick_
	at $CLAW/geoclaw/src/2d/shallow/tick.f:226
#4  0x525304 in amr2
	at $CLAW/geoclaw/src/2d/shallow/amr2.f90:646
#5  0x52873c in main
	at $CLAW/geoclaw/src/2d/shallow/amr2.f90:59
Traceback (most recent call last):
  File "$CLAW/clawutil/src/python/clawutil/runclaw.py", line 249, in runclaw
    proc = subprocess.check_call(cmd_split,
  File "$CONDA_PREFIX/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['$WORKDIR/xgeoclaw']' died with <Signals.SIGFPE: 8>.
```
Here is the setup: [gc_segfpe.zip](https://github.com/clawpack/geoclaw/files/7528364/gc_segfpe.zip)

Would you prefer to have a separate issue for this or continue to discuss this here?

**Comment by mjberger** [comment]
what compiler options and compiler? Both?

 — Marsha

**Comment by tovogt** [comment]
This is with gfortran 7.3.0:
```
FFLAGS='-fopenmp -O0 -W -Wall -fbounds-check -fcheck=all -Wunderflow -fbacktrace -ffpe-trap=invalid,zero,overflow -g'
```
I currently try to reproduce this with `ifort`. The process has been running for 7 hours without failing, but it's not finished yet ...

**Comment by mandli** [comment]
I am running it through my setup with checkpoints turned on.  I will inform if I can reproduce the problem.

**Comment by mandli** [comment]
Oh, and due to the location of the signal I am somewhat suspicious that this may be a related problem.

**Comment by tovogt** [comment]
Unfortunately, I still can't tell whether this also applies to the Intel compiler because the process has been cancelled by our cluster after 24 hours due to a time limit. I had to restart with a longer time limit and that will take a while now...

**Comment by mandli** [comment]
I did finish the run on my laptop, it took nearly 48 hours on an 8 core i9.  Looking at the output it looks like the run is refining everywhere, even after the storm is well inland.  One thing that I am somewhat concerned with is the size of the domain relative to the storm is quite small, which can cause boundary condition problems.

**Comment by tovogt** [comment]
When you say "the size of the domain relative to the storm is quite small", what do you take as a rule of thumb to determine an appropriate domain size?

Currently, I select the domain to accommodate the `storm_radius`, which is orders of magnitude larger than the `max_wind_radius` storm variable. I assumed that most surge dynamics would be concentrated within that area. Would you go for twice the `storm_radius`? Or three times, ...?

My experience from larger domain size is that it will increase runtimes in most cases dramatically, won't change the results in the landfall area of the storm significantly, and will sometimes cause GeoClaw to spend a lot of time computing complex interaction of dynamics with topographic features that are far off the region of interest.

**Comment by mandli** [comment]
I often run with large domains where the storm is at least 2-3 `storm_radius` interior to the domain but lately I have also had `storm_radius` set to a larger number as that parameter is often not provided.  The key here though is to restrict the resolution in most of the domain to be low but keeping the resolution criteria free to refine wither only in the near shore or, even better, only along the track of the storm.

**Comment by tovogt** [comment]
In fact, in this case, enlarging the domain to 2.5 `storm_radius` and enforcing a low resolution outside of a quite narrow window around the storm, does help: There is no segmentation fault and the whole simulations is done on our cluster in less than 25 minutes (previously, it ran for over an hour until the segmentation fault happened).

**Comment by mandli** [comment]
Yes, that's still worrying but I wonder now if it's somehow related to either length of time or amount of memory that the process is requesting.

**Comment by mjberger** [comment]
For me with the ifort compiler it was the memory size. The code died in an unusual place, and was able to run through it when I enlarged the stack size limit.

 — Marsha

**Comment by tovogt** [comment]
Currently, I have this setting:
```
export OMP_STACKSIZE=500M
export GOMP_STACKSIZE=$OMP_STACKSIZE
export KMP_STACKSIZE=$OMP_STACKSIZE

ulimit -t unlimited              # cputime
ulimit -f unlimited              # filesize
ulimit -d unlimited              # datasize
ulimit -s unlimited              # stacksize
ulimit -c unlimited              # coredumpsize
ulimit -v unlimited              # vmemoryuse
ulimit -l unlimited              # memorylocked
```
I will rerun with gfortran and a higher `OMP_STACKSIZE` and see whether it still fails or whether the time of failure changes.

**Comment by mandli** [comment]
I definitely had upwards of 4-5 GB of memory dedicated to the run so the stack size problem may well be the issue.  The variables `GOMP_STACKSIZE` and `KMP_STACKSIZE` I have seen override the `ulimit` command so I am now wondering if that is the real problem.

**Comment by tovogt** [comment]
Sure, in my setup, the threads also have 4 GB of memory each. But isn't the stack size a different thing and will typically not be larger than a few megabytes? On the contrary, I thought that a larger stack size will reduce the amount of available "heap" memory (because memory is reserved for the stack), so that it's also not advisable to increase the stacksize indefinitely.

**Comment by tovogt** [comment]
I can confirm that this runs through with the Intel compiler (even with STACKSIZE set to 500M, as above).

Furthermore, running with gfortran and a larger STACKSIZE (1000M), there won't be a segmentation fault. Still, the run will fail with
```
 **** Too many dt reductions ****
 **** Stopping calculation   ****
```
How is it possible that the behavior of GeoClaw depends on the compiler in a such a fundamental way?

**Comment by mandli** [comment]
Optimization flags between compilers can be very different and at times can lead to incorrect mathematics being produced, enough so that impacts us.  It's unfortunately a bit of a moving target as it does change between versions of compilers.  My best guess as to why the stack size thing is happening is that gfortran is doing stack memory different (and less efficiently) than in ifort.

For the dt reductions I would hazard also that the run is blowing up.  Not sure if you can plot the problem to see.

**Comment by tovogt** [comment]
Yes, maybe let's agree that this issue can be closed once clawpack/amrclaw#272 is merged (thanks again for that!).

**Comment by mandli** [comment]
Great to hear!  As you suggested, once clawpack/amrclaw#272 is merged let's close this and open up something else if needed.
